pkg/ebpf/dns_parser.go

```diff
@@ -41,7 +43,7 @@ func newDNSParser() *dnsParser {
 func (p *dnsParser) ParseInto(data []byte, t *translation) error {
 	err := p.decoder.DecodeLayers(data, &p.layers)
 	if err != nil || p.decoder.Truncated {
-		return nil
+		return errParsing
```

Comment:
shouldn't we simply propagate the returned error upstream, instead of replacing it by `errParsing`?

Reply:
I guess that will require something like:
```
if err != nil {
    return err
}
if p.decoder.Truncated {
   return errTruncated
}
```

Reply:
The reason for replacing `err` by `errParsing` is to allow the callsite to increment `decodingError` only when the error type is `errParsing`. If we propagate the returned error upstream (which can be of different types), then we won't be able to do that kind of filtering. If we don't want to consider `errTruncated` as a `decodingError` then I can do the following (a slightly modified version of the above code snippet):

```go
if err != nil {
    return errParsing
}
if p.decoder.Truncated {
    return errTruncated
}
```

Reply:
As I mentioned in the ticket, in the few cases I was able to get segmentation/fragmentation to happen at the TCP/IP layer, the `decoder.Truncated` bit was **not** set. For example, the code for decoding DNS layer does not set the `truncated` bit when it finds the DNS payload length is greater than the TCP segment: https://github.com/google/gopacket/blob/master/layers/dns.go#L719.

That means separately counting `p.decoder.Truncated` cases don't buy us much since it will **NOT** reflect the actual number of packets we failed to process due to our lack of ability to handled segmentation.

In that cases, I propose we merge all parsing errors and truncated cases into one and report one error (the initial iteration of this PR).

Reply:
I understand your point, thought I still find useful to not override the error types inside the decoder (otherwise, why bother to have different errors?)

If you are thinking on merging them for the purpose of emitting a metric *upstream*, you could replace [this switch statement](https://github.com/DataDog/datadog-agent/pull/4878/files#diff-556e609dcd5f19737208a66195b127d4R125-R130) by something like the following:

```go
switch err {
case errTruncated:
	atomic.AddInt64(&s.truncatedPkts, 1)
default:
	atomic.AddInt64(&s.decodingErrors, 1)
	log.Tracef("error decoding DNS payload: %v", err)
}
```

Regarding the truncation error, I think it's worth keeping the `if p.decoder.Truncated` and trying (on a future PR) to see if you can do something similar to [this](https://github.com/google/gopacket/blob/master/layers/dns.go#L304-L310) on your custom TCP layer somehow.